File: fm-dddrive/src/main/java/io/dddrive/core/property/model/base/PropertyBase.java
package io.dddrive.core.property.model.base;

import io.dddrive.core.property.model.EntityWithProperties;
import io.dddrive.core.property.model.EntityWithPropertiesSPI;
import io.dddrive.core.property.model.Property;

public abstract class PropertyBase<T> implements Property<T> {

	private final EntityWithProperties entity;
	private final String name;

	protected PropertyBase(EntityWithProperties entity, String name) {
		this.entity = entity;
		this.name = name;
	}

	@Override
	public EntityWithProperties getEntity() {
		return this.entity;
	}

	@Override
	public String getRelativePath() {
		String relativePath = ((EntityWithPropertiesSPI) this.entity).getRelativePath();
		return relativePath.isEmpty() ? this.name : relativePath + "." + this.name;
	}

	@Override
	public String getPath() {
		return ((EntityWithPropertiesSPI) this.entity).getPath() + "." + this.name;
	}

	@Override
	public String getName() {
		return this.name;
	}

	@Override
	public boolean isWritable() {
		return !this.getEntity().isFrozen();
	}

}

File: fm-dddrive/src/main/java/io/dddrive/core/property/model/base/PropertyBase.kt
package io.dddrive.core.property.model.base

import io.dddrive.core.property.model.EntityWithProperties
import io.dddrive.core.property.model.EntityWithPropertiesSPI
import io.dddrive.core.property.model.Property

abstract class PropertyBase<T : Any>(
	override val entity: EntityWithProperties,
	override val name: String,
) : Property<T> {

	override val relativePath: String
		get() {
			val relativePath = (this.entity as EntityWithPropertiesSPI).relativePath
			return if (relativePath.isEmpty()) this.name else relativePath + "." + this.name
		}

	override val path: String
		get() = (this.entity as EntityWithPropertiesSPI).path + "." + this.name

	override val isWritable: Boolean
		get() = !this.entity.isFrozen

}
